File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        
        if n == 0: return 0
        
        l = 0
        r = n-1
        res = 0
        leftMax = height[l]
        rightMax = height[r]
        
        while l < r:
            if leftMax < rightMax:
                l += 1
                leftMax = max(leftMax, height[l])
                res += leftMax - height[l]
            else:
                r -= 1
                rightMax = max(rightMax, height[r])
                res += rightMax - height[r]
        
        return res

File: 0042-trapping-rain-water/0042-trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    const n = height.length
    let max = 0
    const left = []
    for (let i=0; i < n; i++) {
        max = Math.max(max, height[i])
        left[i] = max
    }
    max = 0
    const right = []
    for (let i=n-1; i >= 0; i--) {
        max = Math.max(max, height[i])
        right[i] = max
    }
    let res = 0
    for (let i=0; i < n; i++) {
        res += Math.min(left[i], right[i]) - height[i]
    }
    return res
};